### ISSUE TITLE
Refactoring Printer

### ISSUE THREAD

**Issue by Chemaclass** [jenshaase/phel-lang#129]
The current `Phel\Printer` class is a "God Class" with too many responsibilities. Actually 19, one per each conditional inside the `print()` function. See:
```php
public function print($form): string
{
    if ($form instanceof Tuple) {
        return $this->printTuple($form);
    }
    if ($form instanceof Keyword) {
        return ':' . $form->getName();
    }
    if ($form instanceof Symbol) {
        return $form->getName();
    }
   // ... etc
```

### Suggestion 1: Using extract class refactoring

- Create a new Module named `Printer` and there allocate each private function in a class. This way it will be easily testable.
- A similar strategy as we already did for the `Analyzer` module.

### Suggestion2: Using polymorphism

- Instead of having all information about how to print anything inside this GodClass, consider moving the responsibility of printing inside each class. For example. Instead of:
```php
if ($form instanceof Tuple) {
    return $this->printTuple($form);
}
```

Use `$form->print()` when required (usually inside their `__toString()` methods).

In my opinion, the ideal option is the second one, because each form type would be responsible for its own printing thing. BUT there is a problem with this approach, there are `forms` without types, also we are using native scalar types (this means not custom-defined types). This way I don’t find an easy way to do this second approach…

**TL;DR**: I would go for the first approach (Suggestion 1: Using extract class refactoring) as a first iteration to reduce the complexity of this existing class. And in another Issue/PR we could discuss the second option (adding custom-types for all forms).

**Comment by jenshaase** [comment]
Suggestion 1 is the better solution! The Printer must be able to print any PHP Type. This can also be classes of other library.

**Comment by JesusValera** [comment]
Implemented here https://github.com/jenshaase/phel-lang/pull/134 :D
